Joss paper (#150)
NorskRegnesentral/shapr: inst/joss_paper/paper.Rmd

USE: devtools::install_github("benmarwick/wordcountaddin", type = "source", dependencies = TRUE)

for word counting in the Rmarkdwon document (250-1000 words)

-->

# Summary

A common task within machine learning is to train a model which is able to predict an unknown outcome 
(response variable) based on a set of known input variables/features.
When using such models for real life applications, it is often crucial to understand why a certain set of features lead 
to exactly a specific prediction.
Most machine learning models are however so complicated and hard to understand that they are often viewed as 
"black-boxes" producing output when provided some input.

Shapley values (@Shapley53) is a concept from cooperative game theory used to fairly distribute a joint payoff among the
cooperating players. 
@kononenko2010efficient and later @lundberg2017unified proposed to use the Shapley value framework to explain 
predictions by distributing the prediction value on the input features. 

> Unfortunately, established methods and implementations for explaining predictions with Shapley values like Shapley 

Reviewer: ```suggestion
Established methods and implementations for explaining predictions with Shapley values like Shapley 
```

Sampling Values (@vstrumbelj2014explaining), SHAP/Kernel SHAP (@lundberg2017unified), and to some extent TreeSHAP 
(@lundberg2018consistent), assume that the features are independent when approximating the Shapley values for prediction
explanation. 

> This `R`-package implements methodology proposed by @aas2019explaining to explain predictions by accounting for the 

Reviewer: ```suggestion
`shapr` implements the methodology proposed by @aas2019explaining, where predictions are explained while accounting for the 
```

Author: ```suggestion
The `R`-package `shapr` implements the methodology proposed by @aas2019explaining, where predictions are explained while accounting for the 
```

Author: Ok with this @nikolase90 ?

Reviewer: Yep!

dependence between the features, resulting in significantly more accurate approximations to the Shapley values. 

# Implementation

The package implements a variant of the Kernel SHAP (@lundberg2017unified) methodology for efficiently dealing with the 

> combinatorial problem related to Shapley value formula. 

Reviewer: ```suggestion
combinatorial problem related to the Shapley value formula. 
```

The main methodological contribution of @aas2019explaining is three different methods to estimate certain conditional 
expectation quantities, referred to as the `empirical`, `gaussian` and `copula` approach. Additionaly, the user has
the option of combining the three approaches. 
The implementation supports explanation of the following models natively: `stats::lm`, `stats::glm`, `ranger::ranger`, 
`mgcv::gam` and `xgboost::xgboost`. 
Moreover, the package supports explanation of custom models by supplying two simple additional class functions.

The user interface in the package has largely been adopted from the `R`-package `lime` (@limeRpackage). 
The user first sets up the explainability framework for the model to explain with the `shapr` function. 
Then the output from `shapr` is provided to the `explain` function, along with the data to explain the prediction for
and which method should be used to estimate the aforementioned conditional expectations.

> The majority of the code is plain `R`, while the most time consuming parts of the code has been coded in `C++` through 

Reviewer: Cite `R`, `data.table` and `Rcpp`.

Author: Fixed in 16ff339

the `Rcpp` package for computational speed up. 

> Our implementation has some minor methodological improvements to the Kernel SHAP methodology as implemented in the 

Reviewer: ```suggestion
Our implementation has methodological improvements to the Kernel SHAP methodology as implemented in the 
```

`Python` package `shap` (@shapPythonpackage). 
In addition to our package's ability to account for the feature dependence (which the `shap` package does not), 

> basic tests suggest our implementation is about 3-4 times faster. 

Reviewer: This sounds vague.

Author: Indeed, it is vague. I don't think we have the space to say what our methodological improvements really are. Ok with skipping the speedup thing and keeping the rest similar to how it is? @nikolase90

Author: Attempted to fix this in 16ff339/ d107cfa, by deleting/moving the relevant parts. What do you think @nikolase90 ?

Reviewer: :+1:

> For more information about the underlying methodology of the package, we refer to the 

Reviewer: ```suggestion
For a detailed description of the underlying methodology of the package, we refer to the 
```

[paper](https://arxiv.org/abs/1903.10464) (@aas2019explaining).

> For getting started with the package, we refer to the package vignette and introductory examples available at the 

Reviewer: ```suggestion
For getting started with the package, we highly recommend the user to go through the package vignette and introductory examples available at the 
```

Author: ```suggestion
For getting started with the package, we recommend the user to go through the package vignette and introductory examples available at the 
```

`mgcv::gam` and `xgboost::xgboost`/`xgboost::xgb.train`. 
Moreover, the package supports explanation of custom models by supplying two simple additional class functions.

For reference, the package also includes a benchmark implementation of the original (independence assuming) version of
Kernel SHAP (@lundberg2017unified), providing identical results to the "official" Kernel SHAP `Python` package `shap` 
(@shapPythonpackage). 
This allows the user to easily see the effect and importance of accounting for the feature dependence.

The user interface in the package has largely been adopted from the `R`-package `lime` (@limeRpackage). 
The user first sets up the explainability framework for the model to explain with the `shapr` function. 
Then the output from `shapr` is provided to the `explain` function, along with the data to explain the prediction for
and which method should be used to estimate the aforementioned conditional expectations.

The majority of the code is plain `R` (@rCore), while the most time consuming operations has been coded in `C++` 
through the `Rcpp` package (@rcppRpackage) for computational speed up. 

> For RAM efficiency and computational seed up of typical bookeeping operations in `R`, we utilize the `data.table` 

Reviewer: ```suggestion
For RAM efficiency and computational speed up of typical bookeeping operations in `R`, we utilize the `data.table` 
```

<!-- A summary describing the high-level functionality and purpose of the software for a diverse, non-specialist audience. -->
<!-- A clear Statement of Need that illustrates the research purpose of the software. -->
<!-- A list of key references, including to other software addressing related needs. -->
<!-- Mention (if applicable) of any past or ongoing research projects using the software and recent scholarly publications enabled by it. -->
<!-- Acknowledgement of any financial support. -->
<!-- As this short list shows, JOSS papers are only expected to contain a limited set of metadata (see example below), a Statement of Need, Summary, Acknowledgements, and References sections. You can look at an example accepted paper. Given this format, a “full length” paper is not permitted, and software documentation such as API (Application Programming Interface) functionality should not be in the paper and instead should be outlined in the software documentation. 

USE: devtools::install_github("benmarwick/wordcountaddin", type = "source", dependencies = TRUE)

for word counting in the Rmarkdwon document (250-1000 words)

-->

# Summary

> A common task within machine learning is to train a model which is able to predict an unknown outcome 

Reviewer 2: ```suggestion
test
```
